packages/light-client-extension-helpers/src/background/json-rpc-provider.ts

```diff
@@ -0,0 +1,82 @@
+import type * as smoldot from "../smoldot"
+
+import type { JsonRpcConnection } from "@polkadot-api/json-rpc-provider"
+import { getSyncProvider } from "@polkadot-api/json-rpc-provider-proxy"
+
+export type Client = Pick<smoldot.Client, "addChain">
+export type AddChainOptions = smoldot.AddChainOptions
+
+/**
+ * Adds a chain to check if the add chain options are valid and then
+ * subsequently removes it, so that it is not included in smoldot's internal
+ * reference count.
+ *
+ */
+const validateAddChainOptions = async (
+  client: Client,
+  options: AddChainOptions,
+) => {
+  const chain = await client.addChain(options)
+  try {
+    chain.remove()
+  } catch (_) {}
+}
+
+/**
+ * Creates a new JSON RPC provider using a client and options.
+ *
+ * This provider will automatically re-connect to the chain if the
+ * connection is lost.
```

Comment:
Ah so actually `getSyncProvider` is reconnecting behind the scenes by the looks of it already, so all this needs to do is use that and things mostly Just Work, right?

This looks way cleaner/simpler than the Effect stuff; good job simplifying!